Move JAXopt code into NeMoS

bagibence opened the pull request:
- Added JAXopt code in `src/nemos/third_party` keeping the license
- Changed imports in NeMoS and JAXopt to reflect this location
- Added a log of changes to JAXopt's readme
- Adjusted the dependencies removing JAXopt from NeMoS's dependencies but adding JAXopt's dev dependencies
  - [x] Decide how of JAXopt's dev depencies are handled
- Adjusted `tox.ini` to run JAXopt's tests

I explored other locations to put the code (top-level, in `src`, in `src/nemos`, `src/nemos/_solvers`), and this turned out to be the best, e.g. allowing imports as `nemos.third_party.jaxopt`.

Some of JAXopt's tests in `isotonic_test.py` fail for me locally.

As NeMoS only requires a subset of the algorithms implemented in JAXopt, removing the rest would make sense before merging.

Fixes #346

BalzaniEdoardo reviewed (commented):
This looks good to me, but we should edit the readme of the package adding a statement acknowledging that we ported the jaxopt code while we are transitioning to optax/optimistix for our solvers. 

I would also link the readme to the jaxopt apache license.

BalzaniEdoardo reviewed (commented):
TODOs

- [x] Add a job to the CI that runs the jaxopt tests via tox
- [x] Move the test dependencies of jaxopt in a separate "jaxopt-tests" optional deps.
- [ ] Acknowledge the port of `jaxopt` in the readme and add a sentence stating that this is a temporary port while we complete the transition to `optax` and `optimistix`.

billbrod commented:
I agree with all Edoardo's points. I would also update CONTRIBUTING to mention this, linking to the readme description.

This practice of repackaging third-party libraries is known as "vendoring", and if you want to see how another library does this, [pip vendors many libraries](https://github.com/pypa/pip/tree/main/src/pip/_vendor). I'd check out that readme and [init](https://github.com/pypa/pip/blob/main/src/pip/_vendor/__init__.py), poke around and see how they handle attribution etc.

Fortunately for us, pip (like nemos) is [MIT-licensed](https://github.com/pypa/pip?tab=MIT-1-ov-file#readme) and vendors several Apache 2-licensed libraries (like jaxopt), e.g., [requests](https://github.com/pypa/pip/blob/main/src/pip/_vendor/requests/LICENSE). So that's an existence proof for license compatibility.

bagibence commented:
Added `tox_jaxopt` to the CI.

Added a readme in `src/nemos/third_party` and explained why JAXopt is vendored, and linked to it in CONTRIBUTING. Should it be mentioned in the main readme as well? It seemed a bit out of place there for me.

BalzaniEdoardo commented:
I removed the numba jitting in `src/nemos/third_party/jaxopt/_src/isotonic.py` which wasn't working. 

The CI of jaxopt wasn't requiring `numba`, so the bug wasn't detected, but nemos does install it as a pynapple dependecy. We do not use `isotonic.py` at all in our code so that doesn't make a difference.

billbrod reviewed (approved):
This looks like a reasonable way to handle it to me. I just noted a couple of small extra pieces of info that I think would be useful. It's probably also useful to write in an issue / the project where the pieces that will be deleted live, so that we don't forget to delete something.

bagibence commented on src/nemos/third_party/README.md:
```diff
@@ -0,0 +1,13 @@
+## Changes made to `JAXopt`
+- Removed docs, examples, benchmarks
+- Changed import paths to import `jaxopt` from `nemos.third_party`
+- Removed Github workflows (`.github`)
```
Added a line about removing numba.
Added the "Our approach ..." sentence after the longer description of the why and when of vendoring.

bagibence commented on pyproject.toml:
```diff
@@ -23,13 +23,12 @@ classifiers = [
     "pynapple>=0.6.6",
 ]
-
+# jaxopt 0.8.5 is included in src/nemos/third_party/jaxopt
```
Moved it!
